Execute MTA builder in Docker

fwilhe opened the pull request:
**changes:**
- Execute MTA build in a docker container, like other build types too.
- [x] add documentation
- [X] add tests

CCFenner commented on vars/mtaBuild.groovy:
```diff
@@ -13,7 +13,8 @@ import groovy.transform.Field
     'extension',
-    'mtaJarLocation'
+    'mtaJarLocation',
+    'dockerImage'
```
keep the order

CCFenner commented on vars/mtaBuild.groovy:
```diff
@@ -69,10 +70,12 @@ def call(Map parameters = [:]) {
         mtaCall += ' build'
 
-        sh  """#!/bin/bash
+        dockerExecute(script: script, dockerImage: configuration.dockerImage, dockerOptions: configuration.dockerOptions) {
```
`configuration.dockerOptions` is not in the key list or documentation

CCFenner commented on vars/mtaBuild.groovy:
```diff
@@ -69,10 +70,12 @@ def call(Map parameters = [:]) {
-        sh  """#!/bin/bash
+        dockerExecute(script: script, dockerImage: configuration.dockerImage, dockerOptions: configuration.dockerOptions) {
+            sh """#!/bin/bash
             export PATH=./node_modules/.bin:${PATH}
```
correct indent

marcusholl commented:
Was there a test running the mta command inside docker without having the mta toolset installed on the docker host (... on the system hosting Jenkins)? I guess we the checks for java and mta starting with line 35 squints now. These checks are now performed on the docker host, but later we run the mta toolset inside a docker (and not on the environment we checked before). An pipeline run on a system using docker, and _not_ having installed the tools on Jenkins itself (which is completely normal/expected when working with docker containers) should fail now.

I see two options here:

1.) we move the checks inside the closure executed on docker
2.) we remove the checks in total. If that is a good decision depends. We can say: we try to "enforce" running with docker containers. In this case we can say: if you run it in docker it will work, we tested the lib in together with suitable docker images. In this case there is no strong need for checking the environment. The less assumptions we make about the environment the more it is reasonable to keep the checks. I personally think it is fine to rely on a docker landscape. But that is only my personal opinion. Decision about the environment (docker or plain) needs to be done by some steering committee.

fwilhe commented:
I think moving the checks into Docker is a good option. I must admit that I did not really notice the verifications before, and sure it makes no sense to first verify what is installed on the jenkins and then run it in Docker.

CCFenner commented on vars/mtaBuild.groovy:
```diff
@@ -32,47 +35,51 @@ def call(Map parameters = [:]) {
 
-        def java = new ToolDescriptor('Java', 'JAVA_HOME', '', '/bin/', 'java', '1.8.0', '-version 2>&1')
-        java.verify(this, configuration)
+        def mtarFileName = dockerExecute(script: script, dockerImage: configuration.dockerImage, dockerOptions: configuration.dockerOptions) {
```
I guess the `dockerExecute` step is implemented as void.

CCFenner commented:
This is a pretty breaking change, customers would need to build their own docker now.

fwilhe commented:
Customers don't need a docker image, since fallback of the docker step is to execute it locally if no image is configured. Should not be breaking, right?

marcusholl commented:
>This is a pretty breaking change, customers would need to build their own docker now.

To my understanding: no. ```dockerExecute``` makes a fallback in case there is no docker image provided, or in case it cannot be used (e.g. no docker daemon running). In such a case we fallback to local execution (which - of course - requires the corresponding tools to be installed locally).

By the way: In the meantime the mta toolset is available publicly. So we are - in general able to create also a docker image for mta. It can be installed with some snippet like this as we do it already somewhere else:

```
curl --fail \
        --cookie "eula_3_1_agreed=tools.hana.ondemand.com/developer-license-3_1.txt;" \
        --output WHEREEVER_IT_GOES_TO \
https://tools.hana.ondemand.com/additional/mta_archive_builder-1.1.0.jar
```

daniel-kurzynski commented:
> By the way: In the meantime the mta toolset is available publicly.
We should

fwilhe commented:
I think a public available mta builder image is nothing that we can do just now, and I think this should not block this PR since the local fallback is available. I fully support this idea.

Is there anything else with regards to this pull request?

CCFenner commented:
> To my understanding: no. ```dockerExecute``` makes a fallback in case there is no docker image provided, or in case it cannot be used (e.g. no docker daemon running). In such a case we fallback to local execution (which - of course - requires the corresponding tools to be installed locally).

Ok, then we're on the safe side 😉

fwilhe commented:
Hi everyone, since my current task depends on this PR, a timely review is highly appreciated. Thanks 👍 

I addressed all comments so far, if I'm not mistaking.

marcusholl commented:
There is one issue more with dockerizing this: Currently mtaBuild works upon absolute directories (check the ```pwd``` statement(s) in ```mtaBuild```). IMO it is a bad idea to work with absolute directories in a docker use case, since the absolute path of the build root folder outside docker is not necessarily the same like the correponding path inside a docker. Convince me from the opposite.

I'm already working on a refactoring for read yaml in another context (providing project configuration), but this is more related to this as one would believe. In that other context I'm trying to get ride of the ```pwd``` in ```mtaBuild```.

marcusholl commented:
Just out of curiosity: why is it important for another task that we are able to execute this in a docker. I mean docker is cool and the approach of dockerizing all the tools in an approach with a lot of benefits. But why is the fact that mta is executed in a docker a prerequisite for something else?

marcusholl commented:
PR #277 fixes the issue with the absolute directories. ```pwd``` was (mis)used here in order to pass in a temporary directory which was populated with the mta.yaml file. Since we are now able to register the yaml file in the test directly there is no need for the ```pwd``` in the tests.

fwilhe commented:
> Just out of curiosity: why is it important for another task that we are able to execute this in a docker.

For the S/4HANA Cloud SDK pipeline, we don't assume that your Jenkins has anything installed except for Docker. All the work is done in containers, so we can't assume there is a JDK or the MTA builder locally available. This is a design decision we made for various reasons, reproducibility of the environment is one of them. I'm currently working on a PR to enable MTA builds as part of our pipeline, and we make use of the mtaBuild provided by this library. But the version without Docker support is for this reason not usable for us. 

Happy to discuss this in more depth, if you're interested.

marcusholl commented:
> Happy to discuss this in more depth

... aaahh. Just inside mta handling in a broader scope. I thought about another (weird) context outside mta depending on mta somehow required in a docker ...

Wrt dockerization no discussion with me required. I believe that running the tools in docker containers is the right way, since it simplifies a lot of things like lifecycle management, clear setup without clashes in environment variables or installed tools as dependencies etc. pp.

CCFenner commented on vars/mtaBuild.groovy:
```diff
@@ -12,10 +12,13 @@ import groovy.transform.Field
 ]
-@Field Set PARAMETER_KEYS = STEP_CONFIG_KEYS
+@Field Set PARAMETER_KEYS = STEP_CONFIG_KEYS.plus([
+    'dockerOptions'
```
What is the reason why we don't add `dockerOptions` to the `STEP_CONFIG_KEYS`?

marcusholl commented:
should not be merge before the absolute file handling for the artifact that should be uploaded (```pwd```) is fixed in neoDeploy. See PR #277
